tidy3d/components/eme/simulation.py

```diff
@@ -0,0 +1,859 @@
+    )
+
+    monitors: Tuple[annotate_type(EMEMonitorType), ...] = pd.Field(
+        (),
+        title="Monitors",
+        description="Tuple of monitors in the simulation. "
+        "Note: monitor names are used to access data after simulation is run.",
+    )
+
+    boundary_spec: BoundarySpec = pd.Field(
+        BoundarySpec.all_sides(PECBoundary()),
+        title="Boundaries",
+        description="Specification of boundary conditions along each dimension. If ``None``, "
+        "PML boundary conditions are applied on all sides. NOTE: for EME simulations, this "
+        "is required to be PECBoundary on all sides. To capture radiative effects, "
+        "move the boundary farther away from the waveguide in the tangential directions, "
+        "and increase the number of modes. The 'ModeSpec' can also be used to try "
+        "different boundary conditions.",
+    )
+
+    sources: Tuple[None, ...] = pd.Field(
+        (),
+        title="Sources",
+        description="Sources in the simulation. NOTE: sources are not currently supported "
+        "for EME simulations. Instead, the simulation performs full bidirectional "
+        "propagation in the 'port_mode' basis. After running the simulation, "
+        "use 'smatrix_in_basis' to use another set of modes or input field.",
+    )
+
+    grid_spec: GridSpec = pd.Field(
+        GridSpec(),
+        title="Grid Specification",
+        description="Specifications for the simulation grid along each of the three directions. "
+        "This is distinct from 'eme_grid_spec', which defines the 1D EME grid in the "
+        "propagation direction.",
+        validate_default=True,
+    )
+
+    subpixel: bool = pd.Field(
+        True,
+        title="Subpixel Averaging",
+        description="If ``True``, uses subpixel averaging of the permittivity "
+        "based on structure definition, resulting in much higher accuracy for a given grid size.",
+    )
+
+    store_port_modes: bool = pd.Field(
+        True,
+        title="Store Port Modes",
+        description="Whether to store the modes associated with the two ports. "
+        "Required to find scattering matrix in basis besides the computational basis.",
+    )
+
+    port_offsets: Tuple[pd.NonNegativeFloat, pd.NonNegativeFloat] = pd.Field(
+        (0, 0),
+        title="Port Offsets",
+        description="Offsets for the two ports, relative to the simulation bounds "
+        "along the propagation axis.",
+    )
+
+    sweep_spec: Optional[EMESweepSpecType] = pd.Field(
+        None,
+        title="EME Sweep Specification",
+        description="Specification for a parameter sweep to be performed during the EME "
+        "propagation step. The results are stored "
+        "in 'sim_data.smatrix'. Other simulation monitor data is not included in the sweep.",
+    )
+
+    constraint: Literal["passive", "unitary", "none"] = pd.Field(
+        "none",
```

Comment:
Usually, we wouldn't use `"none"` but rather just `None`, i.e. make the constraint optional. If there's no good reason to keep the string I would suggest refactoring in that way.

Reply:
done